src/Avalonia.Styling/Controls/NameScope.cs

```diff
@@ -102,15 +71,29 @@ public void Register(string name, object element)
             else
             {
                 _inner.Add(name, element);
-                Registered?.Invoke(this, new NameScopeEventArgs(name, element));
+                if (_pendingSearches.TryGetValue(name, out var tcs))
+                {
+                    _pendingSearches.Remove(name);
+                    tcs.SetResult(element);
+                }
             }
         }
 
-        /// <summary>
-        /// Finds a named element in the name scope.
-        /// </summary>
-        /// <param name="name">The name.</param>
-        /// <returns>The element, or null if the name was not found.</returns>
+        public ValueTask<object> FindAsync(string name)
+        {
+            var found = Find(name);
+            if (found != null)
+                return new ValueTask<object>(found);
+            if (IsCompleted)
+                return new ValueTask<object>((object)null);
+            if (!_pendingSearches.TryGetValue(name, out var tcs))
+                // We are intentionally running continuations synchronously here
+                _pendingSearches[name] = tcs = new TaskCompletionSource<object>();
```

Comment:
Do we actually have a guarantee that the await-ed continuation will run synchronously? I'd feel more comfortable with all of this if we had our own custom awaitable that we used instead of `ValueTask<object>` and `TaskCompletionSource<object>` here. It looks like the fact that the continuation from the `TaskCompletionSource` runs synchronously is a conditional implementation detail where various factors can change whether or not it is inlined.

Reply:
Done.

I think we'd also want a custom awaitable for our dialogs, clipboard and other stuff, since people are attempting to use `ShowAsync().Result` and wonder why it doesn't work.

Reply:
I agree. If we can help our users avoid trying to do sync-over-async, that'll make our users happier and reduce the number of reported bugs we see from users trying sync-over-async.